crypto/blake2/blake2b.c

```diff
@@ -0,0 +1,310 @@
+
+    v[8]  = blake2b_IV[0];
+    v[9]  = blake2b_IV[1];
+    v[10] = blake2b_IV[2];
+    v[11] = blake2b_IV[3];
+    v[12] = S->t[0] ^ blake2b_IV[4];
+    v[13] = S->t[1] ^ blake2b_IV[5];
+    v[14] = S->f[0] ^ blake2b_IV[6];
+    v[15] = S->f[1] ^ blake2b_IV[7];
+#define G(r,i,a,b,c,d) \
+    do { \
+        a = a + b + m[blake2b_sigma[r][2*i+0]]; \
+        d = rotr64(d ^ a, 32); \
+        c = c + d; \
+        b = rotr64(b ^ c, 24); \
+        a = a + b + m[blake2b_sigma[r][2*i+1]]; \
+        d = rotr64(d ^ a, 16); \
+        c = c + d; \
+        b = rotr64(b ^ c, 63); \
+    } while(0)
+#define ROUND(r)  \
+    do { \
+        G(r,0,v[ 0],v[ 4],v[ 8],v[12]); \
+        G(r,1,v[ 1],v[ 5],v[ 9],v[13]); \
+        G(r,2,v[ 2],v[ 6],v[10],v[14]); \
+        G(r,3,v[ 3],v[ 7],v[11],v[15]); \
+        G(r,4,v[ 0],v[ 5],v[10],v[15]); \
+        G(r,5,v[ 1],v[ 6],v[11],v[12]); \
+        G(r,6,v[ 2],v[ 7],v[ 8],v[13]); \
+        G(r,7,v[ 3],v[ 4],v[ 9],v[14]); \
+    } while(0)
+    ROUND(0);
+    ROUND(1);
+    ROUND(2);
+    ROUND(3);
+    ROUND(4);
+    ROUND(5);
+    ROUND(6);
+    ROUND(7);
+    ROUND(8);
+    ROUND(9);
+    ROUND(10);
+    ROUND(11);
+
+    for(i = 0; i < 8; ++i) {
+        S->h[i] = S->h[i] ^ v[i] ^ v[i + 8];
+    }
+
+#undef G
+#undef ROUND
+}
+
+int BLAKE2b_Update(BLAKE2B_CTX *c, const void *data, size_t datalen) {
+    const uint8_t *in = data;
+    size_t left, fill;
+
+    while(datalen > 0) {
+        left = c->buflen;
+        fill = 2 * BLAKE2B_BLOCKBYTES - left;
+        if(datalen > fill) {
+            memcpy(c->buf + left, in, fill); /* Fill buffer */
+            c->buflen += fill;
+            blake2b_increment_counter(c, BLAKE2B_BLOCKBYTES);
+            blake2b_compress(c, c->buf); /* Compress */
+            /* Shift buffer left */
+            memcpy(c->buf, c->buf + BLAKE2B_BLOCKBYTES, BLAKE2B_BLOCKBYTES);
+            c->buflen -= BLAKE2B_BLOCKBYTES;
+            in += fill;
+            datalen -= fill;
+        } else { /* inlen <= fill */
+            memcpy(c->buf + left, in, datalen);
+            c->buflen += datalen; /* Be lazy, do not compress */
+            in += datalen;
+            datalen -= datalen;
+        }
+    }
+
+    return 1;
+}
+
+int BLAKE2b_Final(unsigned char *md, BLAKE2B_CTX *c)
+{
+    uint8_t buffer[BLAKE2B_OUTBYTES] = {0};
+    int i;
+
+    if(c->buflen > BLAKE2B_BLOCKBYTES) {
+        blake2b_increment_counter(c, BLAKE2B_BLOCKBYTES);
+        blake2b_compress(c, c->buf);
+        c->buflen -= BLAKE2B_BLOCKBYTES;
+        memcpy(c->buf, c->buf + BLAKE2B_BLOCKBYTES, c->buflen);
+    }
+
+    blake2b_increment_counter(c, c->buflen);
+    blake2b_set_lastblock(c);
+    /* Padding */
+    memset(c->buf + c->buflen, 0, 2 * BLAKE2B_BLOCKBYTES - c->buflen);
+    blake2b_compress(c, c->buf);
+
+    /* Output full hash to temp buffer */
+    for(i = 0; i < 8; ++i) {
+        store64(buffer + sizeof(c->h[i]) * i, c->h[i]);
+    }
+
+    memcpy(md, buffer, BLAKE2B_DIGEST_LENGTH);
```

Comment:
On Tue, Mar 8, 2016 at 10:10 AM, Kurt Roeckx <notifications@github.com>
wrote:

> In crypto/blake2/blake2b.c
> <https://github.com/openssl/openssl/pull/566#discussion_r55400511>:
>
> > +        c->buflen -= BLAKE2B_BLOCKBYTES;
> > +        memcpy(c->buf, c->buf + BLAKE2B_BLOCKBYTES, c->buflen);
> > +    }
> > +
> > +    blake2b_increment_counter(c, c->buflen);
> > +    blake2b_set_lastblock(c);
> > +    /* Padding */
> > +    memset(c->buf + c->buflen, 0, 2 * BLAKE2B_BLOCKBYTES - c->buflen);
> > +    blake2b_compress(c, c->buf);
> > +
> > +    /* Output full hash to temp buffer */
> > +    for(i = 0; i < 8; ++i) {
> > +        store64(buffer + sizeof(c->h[i]) * i, c->h[i]);
> > +    }
> > +
> > +    memcpy(md, buffer, BLAKE2B_DIGEST_LENGTH);
>
> I don't see it. Both are character arrays, and then you just do a memcpy()
> from one to the other.
>
You're right.  The buffer is needed in the upstream code because the output
length is a parameter to Final, and store64 might overrun the buffer.
Fixed.